src/main/java/org/team1540/robot2025/subsystems/elevator/Elevator.java

```diff
@@ -0,0 +1,140 @@
+    private final AverageFilter positionFilter = new AverageFilter(10);
+    private double setpointMeters;
+    private final double DEADZONE = 0.2;
+
+    private static boolean hasInstance = false;
+
+    private Elevator(ElevatorIO elevatorIO) {
+        if (hasInstance) throw new IllegalStateException("Instance of elevator already exists");
+        hasInstance = true;
+        this.io = elevatorIO;
+    }
+
+    public static Elevator createReal() {
+        if (Constants.currentMode != Constants.Mode.REAL) {
+            DriverStation.reportWarning("Using real elevator on simulated robot", false);
+        }
+        return new Elevator(new ElevatorIOTalonFX());
+    }
+
+    public static Elevator createSim() {
+        if (Constants.currentMode == Constants.Mode.REAL) {
+            DriverStation.reportWarning("Using simulated elevator on real robot", false);
+        }
+        return new Elevator(new ElevatorIOSim());
+    }
+
+    public static Elevator createDummy() {
+        if (Constants.currentMode == Constants.Mode.REAL) {
+            DriverStation.reportWarning("Using dummy elevator on real robot", false);
+        }
+        return new Elevator(new ElevatorIO() {});
+    }
+
+    @Override
+    public void periodic() {
+        io.updateInputs(inputs);
+        Logger.processInputs("Elevator", inputs);
+
+        if (RobotState.isDisabled()) stop();
+
+        MechanismVisualiser.setElevatorPosition(inputs.positionMeters);
+        positionFilter.add(inputs.positionMeters);
+    }
+
+    public void setElevatorPosition(double positionMeters) {
+        positionMeters = MathUtil.clamp(positionMeters, MIN_HEIGHT, MAX_HEIGHT);
+        setpointMeters = positionMeters;
+        io.setSetpointMeters(setpointMeters);
+    }
+
+    public boolean isAtSetpoint() {
+        return MathUtil.isNear(setpointMeters, positionFilter.getAverage(), POS_ERR_TOLERANCE_METERS) || (inputs.atLowerLimit && setpointMeters <= 0);
+    }
+
+    public void setVoltage(double voltage) {
+        io.setVoltage(voltage);
+    }
+
+    public void stop() {
+        io.setVoltage(0.0);
+    }
+
+    @AutoLogOutput(key = "Elevator/setpoint")
+    public double getSetpoint() {
+        return setpointMeters;
+    }
+
+    public double getPosition() {
+        return inputs.positionMeters;
+    }
+
+    public double getVelocity() {
+        return inputs.velocityMPS;
+    }
+
+    public void setBrakeMode(boolean isBrakeMode) {
+        io.setBrakeMode(isBrakeMode);
+    }
+
+    public boolean getUpperLimit() {
+        return inputs.atUpperLimit;
+    }
+
+    public boolean getLowerLimit() {
+        return inputs.atLowerLimit;
+    }
+
+    public void holdPosition() {
+        setElevatorPosition(inputs.positionMeters);
+    }
+
+    public Command setpointCommand(ElevatorState state) {
+        return Commands.runOnce(() -> setElevatorPosition(state.elevatorHeight), this)
+                .until(this::isAtSetpoint);
+    }
```

Comment:
You might want a few more factories with more termination behavior options, like a non-blocking one that doesn't wait for the elevator to hit the setpoint, and one that runs persistently until externally canceled.

Reply:
Disagree about the non-blocking one, that's something that can and should be handled by the command compositions, not the command factory itself.

More importantly, nothing actually happens when this command ends, so the notion of it ending doesn't really make sense

Reply:
Yes, you can wrap a `setPosition` call in `Commands.runOnce`, but 1) it's more convenient to just call a factory, 2) `setPosition` has a different API, as it currently doesn't take a state enum and takes an explicit position (i suppose this could change), and 3) the  "best practices" dictated by the WPILib people say that interactions with protected state, such as a positional setpoint, should be exposed through command factories, not method calls.

The idea of when the command ends is still important, as it  dictates when the command releases requirements.

Reply:
What I meant with the first part is that you should do a parallel rather than sequential if you don't want it to be blocking